pkg/xds/ingress/dataplane.go

```diff
@@ -91,6 +111,66 @@ func availableServicesEqual(services []*mesh_proto.ZoneIngress_AvailableService,
 	return true
 }
 
+func getMeshGateways(
+	ctx context.Context,
+	rm manager.ResourceManager,
+	dataplanes []*core_mesh.DataplaneResource,
+	meshGateways []*core_mesh.MeshGatewayResource,
+) ([]xds_context.MeshGatewayDataplanes, error) {
+	meshGatewayDataplanes := []xds_context.MeshGatewayDataplanes{}
+
+	meshGatewaysByMesh := map[xds.MeshName][]*core_mesh.MeshGatewayResource{}
+	for _, gateway := range meshGateways {
+		gateways := meshGatewaysByMesh[gateway.GetMeta().GetMesh()]
+		meshGatewaysByMesh[gateway.GetMeta().GetMesh()] = append(gateways, gateway)
+	}
+
+	dataplanesByMesh := map[xds.MeshName][]*core_mesh.DataplaneResource{}
+	for _, dataplane := range dataplanes {
+		if !dataplane.Spec.IsBuiltinGateway() {
+			continue
+		}
+		dataplanes := dataplanesByMesh[dataplane.GetMeta().GetMesh()]
+		dataplanesByMesh[dataplane.GetMeta().GetMesh()] = append(dataplanes, dataplane)
+	}
+
+	for meshName, meshGateways := range meshGatewaysByMesh {
+		dataplanes := dataplanesByMesh[meshName]
+		mesh := core_mesh.NewMeshResource()
+		if err := rm.Get(ctx, mesh, store.GetByKey(meshName, model.NoMesh)); err != nil {
+			return nil, err
+		}
+
+		meshGatewayDataplanes = append(meshGatewayDataplanes, xds_context.MeshGatewayDataplanes{
+			Mesh:       mesh,
+			Gateways:   meshGateways,
+			Dataplanes: dataplanes,
+		})
+	}
+
+	return meshGatewayDataplanes, nil
+}
+
+func getIngressAvailableMeshGateways(mesh *core_mesh.MeshResource, meshGateways []*core_mesh.MeshGatewayResource, dataplanes []*core_mesh.DataplaneResource) []*mesh_proto.ZoneIngress_AvailableService {
+	// We just use this to get the meshGateways to add them to availableservices
+	endpointMap := topology.BuildCrossMeshEndpointMap(
```

Comment:
I might be wrong, but it seems like if you create a separate method to get the meshGateways, you don't need a full `mesh` resource, only its name. This means you don't have to make a request to resourceManager in `getMeshGateways`